### PR TITLE
8339799: Reduce work done in j.l.invoke bytecode generators

### PR DESCRIPTION
**Pull request by cl4es**
Misc improvements to InnerClassLambdaMetafactory and InvokerBytecodeGenerator:
- Define `ClassEntry` early, use it when profitable (there are some warts in the API where using `ce.name(), ce.type()` in a few places means we drop the link to the `ClassDesc` which will then be re-spun in later - should be revisited)
- Use `mh.invokeBasic()` for zero-arg non-capturing lambda constructor calls(!) 
- Narrow name validation to the name passed in
- We were calling `classDesc` twice per factoryType argument, refactored to reuse `argDescs`. 

Interpreter instrumentation sees 1.5% less bytecode executed on netty startup tests, -2.5% on minimal lambda test. Part of the https://bugs.openjdk.org/browse/JDK-8338542 ClassFile API startup work.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8339799](https://bugs.openjdk.org/browse/JDK-8339799): Reduce work done in j.l.invoke bytecode generators (**Sub-task** - P4)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20925/head:pull/20925` \
`$ git checkout pull/20925`

Update a local copy of the PR: \
`$ git checkout pull/20925` \
`$ git pull https://git.openjdk.org/jdk.git pull/20925/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20925`

View PR using the GUI difftool: \
`$ git pr show -t 20925`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20925.diff">https://git.openjdk.org/jdk/pull/20925.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20925#issuecomment-2339296599)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java [hunk 1]
```diff
@@ -97,7 +97,8 @@
     private final String[] argNames;                 // Generated names for the constructor arguments
     private final ClassDesc[] argDescs;              // Type descriptors for the constructor arguments
     private final String lambdaClassName;            // Generated name for the generated class "X$$Lambda$1"
-    private final ClassDesc lambdaClassDesc;         // Type descriptor for the generated class "X$$Lambda$1"
+    private final ConstantPoolBuilder pool = ConstantPoolBuilder.of();
+    private final ClassEntry lambdaClassEntry;       // Class entry for the generated class "X$$Lambda$1"
```
Comment (liach) [inline src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java:101]: Does this improvement go away if remove this change to use explicit class entries? I think I can do some constant pool work to improve ClassEntry retrieval from ClassDesc. Something like #20667, but I will try move the hash to ClassEntry to enforce ClassDesc primacy.

Comment (cl4es) [reply src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java:101]: When experimenting with this all the changes in this PR individually moved the needle in the right direction. If you have doubts about this facet in particular I can split it out and we can re-evaluate it together with #20667.

**Review by liach — APPROVED** [review]
After all, this patch will still be valid with ClassEntry hash updates.

**Comment by cl4es** [comment]
> After all, this patch will still be valid with ClassEntry hash updates.

Thanks for checking and reviewing!

/integrate

**Pull request closed without merging**